File: ios/DoMemory/DoMemory/Modules/Multiplayer/Model/MultiplayerModels.swift
//
//  MultiplayerModels.swift
//  DoMemory
//

import Foundation

enum MultiplayerRoomStatus: String, Codable {
    case waiting
    case ready
    case playing
    case reconnecting
    case finished
    case abandoned
}

enum MultiplayerGameSource: String, Codable {
    case firebase
    case custom
}

struct MultiplayerPlayer: Codable, Identifiable, Hashable {
    var id: String
    var name: String
    var connected: Bool
    var lastSeenAt: TimeInterval
    var score: Int
}

struct MultiplayerCard: Codable, Identifiable, Hashable {
    var id: Int
    var itemId: Int
    var content: String
    var isFaceUp: Bool
    var isMatched: Bool
}

struct MultiplayerCustomGamePayload: Codable, Hashable {
    var title: String
    var category: String
    var items: [String]
    var itemType: String
    var isDoubleItem: Bool
}

struct MultiplayerRoom: Codable, Identifiable, Hashable {
    var id: String
    var code: String
    var status: MultiplayerRoomStatus
    var createdAt: TimeInterval
    var updatedAt: TimeInterval
    var hostId: String
    var guestId: String?
    var players: [String: MultiplayerPlayer]
    var gameSource: MultiplayerGameSource
    var gameId: String
    var gameName: String
    var difficulty: String
    var customGamePayload: MultiplayerCustomGamePayload?
    var currentPlayerId: String?
    var cards: [MultiplayerCard]
    var selectedCardIds: [Int]
    var winnerId: String?
    var disconnectStartedAt: TimeInterval?
    var disconnectPlayerId: String?

    enum CodingKeys: String, CodingKey {
        case id
        case code
        case status
        case createdAt
        case updatedAt
        case hostId
        case guestId
        case players
        case gameSource
        case gameId
        case gameName
        case difficulty
        case customGamePayload
        case currentPlayerId
        case cards
        case selectedCardIds
        case winnerId
        case disconnectStartedAt
        case disconnectPlayerId
    }

    init(
        id: String,
        code: String,
        status: MultiplayerRoomStatus,
        createdAt: TimeInterval,
        updatedAt: TimeInterval,
        hostId: String,
        guestId: String?,
        players: [String: MultiplayerPlayer],
        gameSource: MultiplayerGameSource,
        gameId: String,
        gameName: String,
        difficulty: String,
        customGamePayload: MultiplayerCustomGamePayload?,
        currentPlayerId: String?,
        cards: [MultiplayerCard],
        selectedCardIds: [Int],
        winnerId: String?,
        disconnectStartedAt: TimeInterval?,
        disconnectPlayerId: String?
    ) {
        self.id = id
        self.code = code
        self.status = status
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.hostId = hostId
        self.guestId = guestId
        self.players = players
        self.gameSource = gameSource
        self.gameId = gameId
        self.gameName = gameName
        self.difficulty = difficulty
        self.customGamePayload = customGamePayload
        self.currentPlayerId = currentPlayerId
        self.cards = cards
        self.selectedCardIds = selectedCardIds
        self.winnerId = winnerId
        self.disconnectStartedAt = disconnectStartedAt
        self.disconnectPlayerId = disconnectPlayerId
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        code = try container.decode(String.self, forKey: .code)
        status = try container.decode(MultiplayerRoomStatus.self, forKey: .status)
        createdAt = try container.decode(TimeInterval.self, forKey: .createdAt)
        updatedAt = try container.decode(TimeInterval.self, forKey: .updatedAt)
        hostId = try container.decode(String.self, forKey: .hostId)
        guestId = try container.decodeIfPresent(String.self, forKey: .guestId)
        players = try container.decode([String: MultiplayerPlayer].self, forKey: .players)
        gameSource = try container.decode(MultiplayerGameSource.self, forKey: .gameSource)
        gameId = try container.decode(String.self, forKey: .gameId)
        gameName = try container.decode(String.self, forKey: .gameName)
        difficulty = try container.decode(String.self, forKey: .difficulty)
        customGamePayload = try container.decodeIfPresent(MultiplayerCustomGamePayload.self, forKey: .customGamePayload)
        currentPlayerId = try container.decodeIfPresent(String.self, forKey: .currentPlayerId)
        cards = try container.decodeIfPresent([MultiplayerCard].self, forKey: .cards) ?? []
        selectedCardIds = try container.decodeIfPresent([Int].self, forKey: .selectedCardIds) ?? []
        winnerId = try container.decodeIfPresent(String.self, forKey: .winnerId)
        disconnectStartedAt = try container.decodeIfPresent(TimeInterval.self, forKey: .disconnectStartedAt)
        disconnectPlayerId = try container.decodeIfPresent(String.self, forKey: .disconnectPlayerId)
    }
}

extension MultiplayerRoom {
    var isHostCurrentUser: Bool {
        hostId == MultiplayerService.currentUserID
    }

    var opponent: MultiplayerPlayer? {
        players.values.first { $0.id != MultiplayerService.currentUserID }
    }

    var currentPlayer: MultiplayerPlayer? {
        guard let currentPlayerId else { return nil }
        return players[currentPlayerId]
    }

    var currentUserPlayer: MultiplayerPlayer? {
        guard let id = MultiplayerService.currentUserID else { return nil }
        return players[id]
    }

    var isCurrentUserTurn: Bool {
        currentPlayerId == MultiplayerService.currentUserID
    }

    var allPairsMatched: Bool {
        !cards.isEmpty && cards.allSatisfy(\.isMatched)
    }
}

File: android/app/src/main/java/com/ezequielbrrt/domemory/services/multiplayer/MultiplayerModels.kt
package com.ezequielbrrt.domemory.services.multiplayer

import com.ezequielbrrt.domemory.core.model.Board
import com.ezequielbrrt.domemory.core.model.Card

/** Wire-compatible room values from spec 10.2 and iOS's MultiplayerModels.swift. */
enum class MultiplayerRoomStatus { WAITING, READY, PLAYING, RECONNECTING, FINISHED, ABANDONED }
enum class MultiplayerGameSource { FIREBASE, CUSTOM }

data class MultiplayerPlayer(
    val id: String,
    val name: String,
    val connected: Boolean = true,
    val lastSeenAt: Long,
    val score: Int = 0,
    // Spec 10.2/10.4: per-player mutual ready-check flag, false until explicitly set true.
    // Reset to false on both players whenever the host (re)picks a game — see
    // MultiplayerService.selectGame.
    val isReady: Boolean = false,
)

data class MultiplayerCustomGamePayload(
    val title: String,
    val category: String,
    val items: List<String>,
    val itemType: String,
    val isDoubleItem: Boolean,
) {
    fun asBoard(id: String, difficulty: String?) = Board(
        id = id, name = title, category = category, difficulty = difficulty,
        items = items, itemType = itemType, isDoubleItem = isDoubleItem,
    )

    companion object {
        fun from(board: Board) = MultiplayerCustomGamePayload(
            title = board.name, category = board.category, items = board.items,
            itemType = board.itemType, isDoubleItem = board.isDoubleItem,
        )
    }
}

/** A room card intentionally only has the persistent fields of [Card]. */
data class MultiplayerCard(
    val id: Int,
    val itemId: Int,
    val content: String,
    val isFaceUp: Boolean = false,
    val isMatched: Boolean = false,
) {
    fun asCard() = Card(id, itemId, content, isFaceUp, isMatched)
    companion object { fun from(card: Card) = MultiplayerCard(card.id, card.itemId, card.content, card.isFaceUp, card.isMatched) }
}

/** Stable card serialization shared by both native clients; do not shuffle this list. */
object MultiplayerCards {
    fun from(board: Board): List<MultiplayerCard> = if (board.isDoubleItem) {
        board.items.flatMapIndexed { itemId, content ->
            listOf(
                MultiplayerCard(id = itemId * 2, itemId = itemId, content = content),
                MultiplayerCard(id = itemId * 2 + 1, itemId = itemId, content = content),
            )
        }
    } else {
        board.items.mapIndexed { index, content ->
            MultiplayerCard(id = index, itemId = if (index % 2 == 0) index else index - 1, content = content)
        }
    }
}

data class MultiplayerRoom(
    val id: String,
    val code: String,
    val status: MultiplayerRoomStatus,
    val createdAt: Long,
    val updatedAt: Long,
    val hostId: String,
    val guestId: String? = null,
    val players: Map<String, MultiplayerPlayer>,
    val gameSource: MultiplayerGameSource,
    val gameId: String,
    val gameName: String,
    val difficulty: String? = null,
    val customGamePayload: MultiplayerCustomGamePayload? = null,
    val currentPlayerId: String? = null,
    val cards: List<MultiplayerCard> = emptyList(),
    val selectedCardIds: List<Int> = emptyList(),
    val winnerId: String? = null,
    val disconnectStartedAt: Long? = null,
    val disconnectPlayerId: String? = null,
) {
    val allPairsMatched get() = cards.isNotEmpty() && cards.all { it.isMatched }
    fun isTurnFor(playerId: String) = status == MultiplayerRoomStatus.PLAYING && currentPlayerId == playerId

    /** Spec 10.2: `gameId == ""` is the wire sentinel for "no game chosen yet" — a room only
     * ever has an empty [gameId] in WAITING/READY, never once [status] reaches PLAYING. */
    val hasSelectedGame get() = gameId.isNotEmpty()

    /**
     * Spec 10.4 step 5: the pure precondition for the automatic, host-only start —
     * `status == ready`, a game picked, exactly two players present, and every player ready.
     * Deliberately doesn't check `hostId == caller` — that half of the guard belongs to
     * whoever is *acting* (see [MultiplayerService.start]'s transaction and
     * `MultiplayerViewModel`'s auto-start trigger), not to the room's own data.
     */
    val readyToAutoStart get() =
        status == MultiplayerRoomStatus.READY &&
            hasSelectedGame &&
            players.size == 2 &&
            players.values.all { it.isReady }

    /**
     * Whether tapping a card right now would actually flip one — mirrors iOS's
     * `MultiplayerRoomViewModel.isInteractionEnabled`. Checked locally (no round trip to
     * Firebase) so the optimistic `CARD_FLIP` haptic a tap fires immediately — before the
     * transaction that actually moves the card even lands — matches what a legal tap would
     * do: it's this player's turn, the room is still playing, and no already-selected pair
     * is sitting there waiting to clear. Takes an `isCurrentUser` predicate rather than a raw
     * id, the same shape [isTurnFor] would need if it also had to resolve "me" from the
     * caller's auth session rather than being handed an id directly.
     */
    fun canFlipNow(isCurrentUser: (String) -> Boolean): Boolean =
        status == MultiplayerRoomStatus.PLAYING &&
            currentPlayerId?.let(isCurrentUser) == true &&
            selectedCardIds.size < 2
}
